rosbag2_cpp/src/rosbag2_cpp/clocks/time_controller_clock.cpp

```diff
@@ -126,4 +157,34 @@ double TimeControllerClock::get_rate() const
   return impl_->rate;
 }
 
+void TimeControllerClock::pause()
+{
+  std::lock_guard<std::mutex> lock(impl_->state_mutex);
+  if (impl_->paused) {
+    return;
+  }
+  // Note: needs to not be paused when taking snapshot, otherwise it will use last ros ref
+  impl_->snapshot();
+  impl_->paused = true;
+  impl_->cv.notify_all();
+}
+
+void TimeControllerClock::resume()
+{
+  std::lock_guard<std::mutex> lock(impl_->state_mutex);
+  if (!impl_->paused) {
+    return;
+  }
+  // Note: needs to not be paused when taking snapshot, otherwise it will use last ros ref
+  impl_->paused = false;
```

Comment:
I dont think this works correctly.   Lets say i have check this clock 1 a second.  '

with no pauses i would expect to get: `1,2,3,4,5,....,12,13` as results.

if i put start pausing at 4 and lift the pause at 10 (6 second pause). i would expect the outcome from the clock to look like this: `1,2,3,4,4,4,4,4,4,4,6,7,8`

pretty sure the output i'm going to get is `1,2,3,4,4,4,4,4,4,4,11,12`

Lets just assume the inital set up `ref.ros = 1` ,` ref.steady = 1`.  if i call now at T=1 I get 1 + 1-1) = 1.
if i call now() at T=4. i get 1 + 4-1 = 4.  still ok.  

Than i hit pause.
ros_time = 4
steady_time = 4

now at t=5 i get 4 because the pause flag now() return 4, and will until i unpause

at T=10 i call resume
ref.ros = 10 //(ref.ros + 10 - ref.steady)
ref.steady = 10.   ref.ros should be 4 not 10.

Reply:
Thanks for this! Opened https://github.com/ros2/rosbag2/pull/730 to fix